File: rama/memoria.py
"""Memoria persistente de Rama: lo que aprende y lo que recuerda de vos."""

import json
import os
import tempfile
from collections import deque
from pathlib import Path
from typing import Any

RUTA_DATOS = Path(__file__).resolve().parent.parent / "data"
ARCHIVO_APRENDIDO = RUTA_DATOS / "aprendido.json"
MAX_HISTORIAL = 40


class Memoria:
    """Guarda hechos aprendidos y el estado de la conversación en curso.

    Lo aprendido se persiste en disco; el historial vive sólo en la sesión.
    """

    def __init__(self, archivo: Path | str | None = None) -> None:
        self.archivo = Path(archivo) if archivo else ARCHIVO_APRENDIDO
        self.aprendido: list[dict[str, Any]] = []
        self.perfil: dict[str, Any] = {}
        self.historial: deque[tuple[str, str]] = deque(maxlen=MAX_HISTORIAL)
        self.ultima_pregunta_sin_respuesta: str | None = None
        self.cargar()

    # ---------------------------------------------------------------- disco

    def cargar(self) -> None:
        if not self.archivo.exists():
            return
        try:
            datos = json.loads(self.archivo.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            # Un archivo corrupto no debe impedir que Rama arranque.
            return
        self.aprendido = datos.get("aprendido", [])
        self.perfil = datos.get("perfil", {})

    def guardar(self) -> None:
        self.archivo.parent.mkdir(parents=True, exist_ok=True)
        datos = {"aprendido": self.aprendido, "perfil": self.perfil}
        # Escritura atómica: nunca dejamos el archivo a medio escribir.
        fd, temporal = tempfile.mkstemp(dir=self.archivo.parent, suffix=".tmp")
        try:
            with os.fdopen(fd, "w", encoding="utf-8") as f:
                json.dump(datos, f, ensure_ascii=False, indent=2)
            os.replace(temporal, self.archivo)
        except BaseException:
            Path(temporal).unlink(missing_ok=True)
            raise

    # ------------------------------------------------------------ contenido

    def aprender(self, pregunta: str, respuesta: str) -> None:
        """Registra un par pregunta/respuesta, actualizando si ya existía."""
        pregunta, respuesta = pregunta.strip(), respuesta.strip()
        for hecho in self.aprendido:
            if hecho["pregunta"].lower() == pregunta.lower():
                hecho["respuesta"] = respuesta
                self.guardar()
                return
        self.aprendido.append({"pregunta": pregunta, "respuesta": respuesta})
        self.guardar()

    def olvidar(self, pregunta: str) -> bool:
        antes = len(self.aprendido)
        objetivo = pregunta.strip().lower()
        self.aprendido = [
            h for h in self.aprendido if h["pregunta"].lower() != objetivo
        ]
        if len(self.aprendido) != antes:
            self.guardar()
            return True
        return False

    def recordar(self, clave: str, valor: Any) -> None:
        self.perfil[clave] = valor
        self.guardar()

    def recuerdo(self, clave: str, defecto: Any = None) -> Any:
        return self.perfil.get(clave, defecto)

    def registrar_turno(self, quien: str, texto: str) -> None:
        self.historial.append((quien, texto))

File: android/app/src/main/java/ar/rama/ai/motor/Memoria.kt
package ar.rama.ai.motor

import java.io.File
import org.json.JSONArray
import org.json.JSONObject

/** Dónde persiste Rama lo que aprende. Se abstrae para poder testear sin disco. */
interface Almacen {
    fun leer(): String?
    fun escribir(contenido: String)
}

/** Almacén volátil, para tests. */
class AlmacenEnMemoria(private var contenido: String? = null) : Almacen {
    override fun leer(): String? = contenido
    override fun escribir(contenido: String) { this.contenido = contenido }
}

/** Almacén en disco con escritura atómica: nunca queda un archivo a medias. */
class AlmacenArchivo(private val archivo: File) : Almacen {
    override fun leer(): String? = try {
        if (archivo.exists()) archivo.readText() else null
    } catch (e: Exception) {
        null
    }

    override fun escribir(contenido: String) {
        archivo.parentFile?.mkdirs()
        val temporal = File(archivo.parentFile, archivo.name + ".tmp")
        temporal.writeText(contenido)
        if (!temporal.renameTo(archivo)) {
            archivo.writeText(contenido)
            temporal.delete()
        }
    }
}

/** Un par pregunta/respuesta que le enseñó el usuario. */
data class Hecho(val pregunta: String, val respuesta: String)

/**
 * Guarda hechos aprendidos y el estado de la conversación en curso.
 * Lo aprendido se persiste; el historial vive sólo en la sesión.
 */
class Memoria(private val almacen: Almacen = AlmacenEnMemoria()) {

    val aprendido = mutableListOf<Hecho>()
    private val perfil = HashMap<String, String>()
    val historial = ArrayDeque<Pair<String, String>>()
    var ultimaPreguntaSinRespuesta: String? = null

    init {
        cargar()
    }

    private fun cargar() {
        val crudo = almacen.leer() ?: return
        try {
            val raiz = JSONObject(crudo)
            val hechos = raiz.optJSONArray("aprendido") ?: JSONArray()
            for (i in 0 until hechos.length()) {
                val h = hechos.getJSONObject(i)
                aprendido.add(Hecho(h.getString("pregunta"), h.getString("respuesta")))
            }
            val guardado = raiz.optJSONObject("perfil")
            if (guardado != null) {
                for (clave in guardado.keys()) perfil[clave] = guardado.getString(clave)
            }
        } catch (e: Exception) {
            // Una memoria corrupta no debe impedir que Rama arranque.
            aprendido.clear()
            perfil.clear()
        }
    }

    private fun guardar() {
        val hechos = JSONArray()
        for (h in aprendido) {
            hechos.put(JSONObject().put("pregunta", h.pregunta).put("respuesta", h.respuesta))
        }
        val raiz = JSONObject()
            .put("aprendido", hechos)
            .put("perfil", JSONObject(perfil.toMap()))
        almacen.escribir(raiz.toString(2))
    }

    /** Registra un par pregunta/respuesta, actualizando si ya existía. */
    fun aprender(pregunta: String, respuesta: String) {
        val p = pregunta.trim()
        val r = respuesta.trim()
        val indice = aprendido.indexOfFirst { it.pregunta.equals(p, ignoreCase = true) }
        if (indice >= 0) aprendido[indice] = Hecho(aprendido[indice].pregunta, r)
        else aprendido.add(Hecho(p, r))
        guardar()
    }

    fun olvidar(pregunta: String): Boolean {
        val objetivo = pregunta.trim()
        val quedaron = aprendido.filterNot { it.pregunta.equals(objetivo, ignoreCase = true) }
        if (quedaron.size == aprendido.size) return false
        aprendido.clear()
        aprendido.addAll(quedaron)
        guardar()
        return true
    }

    fun recordar(clave: String, valor: String) {
        perfil[clave] = valor
        guardar()
    }

    fun recuerdo(clave: String): String? = perfil[clave]

    fun registrarTurno(quien: String, texto: String) {
        historial.addLast(quien to texto)
        while (historial.size > MAX_HISTORIAL) historial.removeFirst()
    }

    companion object {
        const val MAX_HISTORIAL = 40
    }
}
